tsl/src/compression/compression.c

```diff
@@ -323,14 +346,142 @@ compress_chunk(Oid in_table, Oid out_table, const ColumnCompressionInfo **column
 
 	TupleDesc in_desc = RelationGetDescr(in_rel);
 	TupleDesc out_desc = RelationGetDescr(out_rel);
+	in_rel_index_oids = RelationGetIndexList(in_rel);
+	int i = 0;
+	/* Before calling row compressor relation should be segmented and sorted as per 
+	 * compress_segmentby and compress_orderby column/s configured in ColumnCompressionInfo.
+	 * Cost of sorting can be mitigated if we find an existing BTREE index defined for
+	 * uncompressed chunk otherwise expensive tuplesort will come into play.
+	 *
+	 * The following code is trying to find an existing index that
+	 * matches the ColumnCompressionInfo so that we can skip sequential scan and 
+	 * tuplesort.
+	 *
+	 * Matching Criteria for Each IndexAtt[i] and ColumnCompressionInfo Keys[i]
+	 * ========================================================================
+	 * a) Index attnum must match with ColumnCompressionInfo Key {keys[i]}.
+	 * b) Index attOption(ASC/DESC and NULL_FIRST) can be mapped with ColumnCompressionInfo
+	 * orderby_asc and null_first.
+	 *
+	 * BTREE Indexes Ordering
+	 * =====================
+	 * a) ASC[Null_Last] ==> [1]->[2]->NULL
+	 * b) [Null_First]ASC ==> NULL->[1]->[2]
+	 * c) DSC[Null_Last]  ==> [2]->[1]->NULL
+	 * d) [Null_First]DSC ==> NULL->[2]->[1]
+	 */
+	if (ts_guc_enable_compression_indexscan)
+	{
+		foreach (lc, in_rel_index_oids)
+		{
+			Oid index_oid = lfirst_oid(lc);
+			Relation index_rel = index_open(index_oid, AccessShareLock);
+			IndexInfo *index_info = BuildIndexInfo(index_rel);
+			int previous_direction = NoMovementScanDirection;
+			int current_direction = NoMovementScanDirection;
 
-	Tuplesortstate *sorted_rel = compress_chunk_sort_relation(in_rel, n_keys, keys);
+			if (n_keys <= index_info->ii_NumIndexKeyAttrs && index_info->ii_Am == BTREE_AM_OID)
+			{
+				for (i = 0; i < n_keys; i++)
+				{
+					int16 att_num = get_attnum(in_table, NameStr(keys[i]->attname));
+
+					int16 option = index_rel->rd_indoption[i];
+					bool index_orderby_asc = ((option & INDOPTION_DESC) == 0);
+					bool index_null_first = ((option & INDOPTION_NULLS_FIRST) != 0);
+					bool is_orderby_asc =
+						COMPRESSIONCOL_IS_SEGMENT_BY(keys[i]) ? true : keys[i]->orderby_asc;
+					bool is_null_first =
+						COMPRESSIONCOL_IS_SEGMENT_BY(keys[i]) ? false : keys[i]->orderby_nullsfirst;
+
+					if (att_num > 0 && index_info->ii_IndexAttrNumbers[i] == att_num)
```

Comment:
Attnum can't be zero here, right? The orderby/segmentby keys must be present in the table. We should check this with an assertion.